Feature watchdog v1 rb

rajkan01 opened the pull request:
Watchdog refactoring to multithreaded thread.
    -Added API to register muliple threads to watchdog drivers
    -Watchdog timeout reconfigures everytime whenever new register thread with longer timeout period
    -New APIs for watchdog
     wd_register(const osThreadId_t tid, const uint32_t timeout) to register to watchdog
     wd_unregister(const osThreadId_t tid) to unregister to watchdog
     kick(const osThreadId_t tid) to refresh the watchdog

<!--
    Required
    Add here detailed changes summary, testing results, dependencies
    Good example: https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html (Pull request template)
-->


### Pull request type

<!--
    Required
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front
    of them would change the meaning incorrectly. The only changes to be made are to add a description text under the
    description heading and to add a 'x' to the correct box.
-->
    [ ] Fix
    [x] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

### Reviewers
@mbed-os-hal

<!--
    Optional
    Request additional reviewers with @username
-->

### Release Notes

<!--
    Optional
    In case of breaking changes, functionality changes or refactors, please add release notes here. 
    For more information, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html#pull-request-types).
-->

Commit:
Add Watchdog HAL API specification headers

HAL watchdog functionality will be implemented as two separate APIs. The reset
reason API allows a user to detect the last system reset reason to identify if
a Watchdog was triggered. The Watchdog API allows configuring and updating
Watchdog timers on all boards. This commit defines the headers.

Commit:
Add K64F reset reason reference implementation

Commit:
Add simple watchdog reference implementation for K64F

- Add preprocessor guard to watchdog api that errors if the reset reason api is not also implemented
- Add RESET_REASON and WATCHDOG to K64F targets.json
- Add watchdog reference implementation

Commit:
Fix watchdog API issues

- Fix typo in module comment
- Redefine the default system behaviour in sleep mode
- Guard K64F enableWait flag
- Remove bit shifts from reset reason enum

fkjagodzinski committed:
Tests: HAL API: Watchdog: Add tests

Commit:
Add Reset Reason platform API

Commit:
Add reset reason reference implementation STM32

Commit:
Add function to fetch platform specific reset reason register values

Commit:
Add independent watchdog reference implementation for STM32

Commit:
Amend reset reason driver API
- Change API to match C++ API throughout drivers
- Amend HAL API documentation to be more specific

Commit:
Add Watchdog driver API

fkjagodzinski committed:
Tests: HAL API: Reset_reason: Add tests

Commit:
Remove window and sleep mode options for watchdog API

fkjagodzinski committed:
Tests: HAL API: Watchdog: Add dev reset tests

Commit:
Add missing license headers

Commit:
Move watchdog parameter validation into the driver layer

Commit:
Fix Watchdog::stop on K64F target

WDOG_Disable will suspend the watchdog until a reset is triggered instead of stopping it altogether. Deinit will disable it until it is reinitialized.

fkjagodzinski committed:
K64F: watchdog HAL: Fix max_timeout.

Limit MAX_TIMEOUT_MS to UINT32_MAX.
Improve MAX_TIMEOUT_MS handling.
Fix hal_watchdog_get_platform_features() returning
.max_timeout = 0xfffffff8 instead of 0xffffffff.

fkjagodzinski committed:
Tests: HAL API: Watchdog: Add time accuracy tests

Commit:
Add RESET_REASON and WATCHDOG definitions to hexiwear build target

Commit:
Rename watchdog.c -> watchdog_api.c to prevent name collision with Platform API

fkjagodzinski committed:
Tests: Drivers: Reset_reason: Add tests

fkjagodzinski committed:
Tests: Drivers: Watchdog: Add tests

bulislaw committed:
Add definition of reset reason and watchdog doxy groups

fkjagodzinski committed:
STM32F4: watchdog HAL: Fix issues found with tests

Fix WATCHDOG_STATUS_INVALID_ARGUMENT for timeout values from 1 ms to
407 ms (fix inability to set prescaler bits to zero).
Fix timeout <-> IWDG registers conversions.
Fix read & write access to IWDG_PR and IWDG_RLR registers.
Fix LSI RC frequency setting.
Limit MAX_TIMEOUT_MS to UINT32_MAX.

fkjagodzinski committed:
K64F: watchdog HAL: Fix init() and stop()

Added a missing wait for the WCT window end. Without it, consecutive
init() or stop() calls were ignored.

fkjagodzinski committed:
STM: HAL: Reset_reason: Correct return values

bulislaw committed:
Add RESET_REASON and WATCHDOG to doxygen options

bulislaw committed:
Document reset reason enum

fkjagodzinski committed:
Tests: HAL: Watchdog: Fix timing accuracy test

Relax time measurement margins for UBLOX_EVK_ODIN_W2.

bulislaw committed:
Disable Watchdog on Odin as it fails intermittently

jeromecoutant committed:
STM32 Watchdog : move API file to STM family level

Commit:
Add Watchdog and ResetReason headers to mbed.h

Adding headers to mbed.h will prevent you having to manually include
them when using the drivers in an application.

stevew817 committed:
Add support for watchdog on Silicon Labs devices

Commit:
Add text to watchdog documentation indicating that it will continue to run in sleep modes

jeromecoutant committed:
STM32 WATCHDOG : compilation issue with typed define

stevew817 committed:
Add implementation of reset reason on Silicon Labs parts

jeromecoutant committed:
STM32 WATCHDOG : update STM32L0 HAL_IWDG_Init to a newest version

jeromecoutant committed:
STM32 WATCHDOG : use ST HAL in order to make code commun for all STM32

Commit:
NCS36510: Added watchdog implementation.

jeromecoutant committed:
STM32 WATCHDOG : increase timeout value

Commit:
fix bugs introduced on rebase

ccli8 committed:
[Nuvoton] Support reset reason

ccli8 committed:
[Nuvoton] Support watchdog timer

ccli8 committed:
[NUC472] Fix WDT driver in BSP

ccli8 committed:
Enlarge wait time for flushing serial buffer in Greentea test code

Original wait time is not enough for UART to flush out before deepsleep/reset on some targets.

ccli8 committed:
Fix NUMAKER_PFM_NANO130 cannot pass mbed_hal/watchdog_reset test

On NUMAKER_PFM_NANO130 target, WDT's clock source is fixed to LIRC, which is much
less accurate than other targets. Enlarge delta define to pass this test.

Commit:
Added reset_reason feature for TMPM066 & TMPM46B

Commit:
Added reset_reason feature for TMPM3H6 & TMPM4G9

Commit:
test(api-watchdog): Fix errors in compiling watchdog greentea tests

The latest rebase of the watchdog feature branch introduced errors in compiling
watchdog tests due to missing headers.

- Watchdog HAL API test
  Include missing header files to main.cpp (mbed_wait_api.h, stdlib.h).

- Watchdog HAL API timing test
  Include missing header files to main.cpp (us_ticker_api.h).

- Watchdog Driver API test
  Include missing header files to main.cpp (mbed_wait_api.h, stdlib.h).
  Inject mbed namespace into main.cpp to fix Watchdog name resolution.

Commit:
fix(hal-watchdog): Add feature guards to K64F watchdog implementation

Commit:
Update license headers

Commit:
Apply new astyle config style

Commit:
Fix more astyle failures

fkjagodzinski committed:
Tests: Watchdog: Fix config access in the host script

Move the calls to get_config_item() to the setup() method. The config is
not available in the initializer yet.

fkjagodzinski committed:
Tests: Watchdog: Update the reset-to-sync delay

Change the config parameter used as a delay before sending the sync
packet after the device reset in watchdog and reset_reason tests. Use
'forced_reset_timeout' instead of 'program_cycle_s'.

Commit:
Added reset_reason feature for TMPM3HQ

Commit:
Edit ResetReason_tests.h

Make minor grammar edits.

fkjagodzinski committed:
Docs: Watchdog: Update code comments

* capitalize driver class name,
* reword test docs for the driver & HAL,
* capitalize Mbed name,
* reword the comments explaining the SERIAL_FLUSH_TIME_MS macro.

Commit:
small grammatical changes

Not sure if the formatting's still alright on these.

Commit:
Capitalizing Watchdog

Commit:
capitalizing class name

Hope this is alright.

Commit:
fix: Remove excess comma from device_has key in targets.json

Commit:
fix(hal-watchdog): Remove 'RESET_REASON' flags from unsupported devices

Commit:
fix(hal-watchdog): Add 'RESET_REASON' flags to supported devices

Commit:
fix(hal-watchdog): Guard TMPM 'ResetReason' code behind feature flag

rajkan01 committed:
Watchdog refactoring to multithreaded thread.
-Added API to register muliple threads to watchdog drivers
-Watchdog timeout reconfigures everytime whenever new register thread with longer timeout period
-New APIs for watchdog
 wd_register(const osThreadId_t tid, const uint32_t timeout) to register to watchdog
 wd_unregister(const osThreadId_t tid) to unregister to watchdog
 kick(const osThreadId_t tid) to refresh the watchdog

rajkan01 committed:
HwWatchdog role :
    - mbed_watchdog_mgr has interface name mbed_wdog_manager_start(),mbed_wdog_manager_stop(),mbed_wdog_manager_kick()
    - HwWatchdog is going to attach with LowPowerTIcker for periodic callback functionality
    - mbed_wdog_manager_start() will either get start either by BL/RTOS Aps,it reads the timeout value specified via macro and macro gets defined in target.json file.
    - mbed_wdog_manager_start() internally configure below HAL hw watchdog with timeout specified via target.json
    - mbed_wdog_manager_start() internally divide the timeout(specified in target.json) by the 2 and attach LowPowerTicker with periodic callback of hw_kick()
    - mbed_wdog_manager_start() internally create one instance of sw watchdog class,to access the static list data structure of sw watchdog class
    - mbed_wdog_manager_kick() function periodically get called and refresh the hw watchdog to avoid watchdog reset
    - converted C++ code into C based APIs
    - added boolean to control watchdog start and stop
    - Added detach from ticker on stop API

rajkan01 committed:
SW Watchdog role:
        -SW watchdog has interface name start(),stop(),kick()	Sw watchdog internally has static list and shared across multiple instance of SW watchdog
	- Sw watchdog initialize timeout value,unique string via constructor whenever threads created sw watchdog object
	-Threads make sure pass proper timeout value,Unique string while creating the instance.
	-start() called by components(BLE,WIFI etc.,),it adds the entry into static list with few details current count ,etc.,
	-kick() called by registered components(BLE,WIFI etc.) to reset current count to zero.
        -is_alive - interface API to mbed_watchdog_manager
        -implementation optimization

rajkan01 committed:
HW Watchdog and SW watchdog interface
  - Added Hw watchdog periodic kick calls software watchdog is_alive check

rajkan01 committed:
Hw Watchdog UnitTest cases
    -Added the unit test case for testing Hw watchdog
    -Added the supported stubs files

rajkan01 committed:
SW watchdog unittest cases
  -Added the mock class function to mock mbed_assert_internal
  -Added the unit test case to test start,kick,stop
  -Modified the interface api name from is_alive to process
  -added the unit test cases for process

fkjagodzinski committed:
Tests: Watchdog: Remove outdated driver tests

The Watchdog driver has been updated making these tests incompatible
with the new API. The Watchdog driver functionality is tested with the
unit tests located in UNITTESTS/drivers/Watchdog/ .

fkjagodzinski committed:
Tests: Watchdog: Add Watchdog Manager tests

rajkan01 committed:
fix: Fixing rebased created bugs

rajkan01 committed:
update:Remove DEVICE_WATCHDOG and added RESET_RESON in stm32 family

rajkan01 committed:
doxy spell check fix for watchdog

rajkan01 committed:
fix for build failure issues

rajkan01 committed:
Revert "fix for build failure issues"

This reverts commit 3ce1d4940cb7fd9d74ca2be75ed08b76854ff534.

rajkan01 committed:
Removed RESET_REASON for NRF52832,NRF52840 family and Removed WATCHDOG for H743ZI board

rajkan01 committed:
fix for Jenkin build issue

rajkan01 committed:
Trailing space issue fix

rajkan01 committed:
Added the SPDX identifier

rajkan01 committed:
Removed RESET_REASON for NRF51_DK board

rajkan01 committed:
Fix for the Ci build issue

rajkan01 committed:
fix for spell check and header file

rajkan01 committed:
fix for astyle

0xc0170 commented on drivers/Watchdog.h:
```diff
@@ -20,21 +20,22 @@
+#include "mbed_error.h"
+#include "platform/mbed_critical.h"
+#include "platform/mbed_power_mgmt.h"
+#include "mbed_assert.h"
```
one empty line between includes and namespace as it was

0xc0170 commented on UNITTESTS/target_h/platform/mbed_power_mgmt.h:
```diff
@@ -15,3 +15,12 @@
+ * @note Does not affect the debug sub-system
+ */
+extern void mock_system_reset();
+MBED_NORETURN static inline void system_reset(void)
```
here also empty line between extern and function def

0xc0170 commented on UNITTESTS/platform/mbed_watchdog_mgr/test_mbed_watchdog_mgr.cpp:
```diff
@@ -0,0 +1,70 @@
+// AStyle ignored as the definition is not clear due to preprocessor usage
+// *INDENT-OFF*
+// AStyle ignored as the definition is not clear due to preprocessor usage
+// *INDENT-OFF*
```
remove duplication (line 25 and 23 are the same)

donatieng commented on UNITTESTS/platform/mbed_watchdog_mgr/Watchdog.cpp:
```diff
@@ -0,0 +1,57 @@
+
+Watchdog *Watchdog::_first;
+
+Watchdog::Watchdog(uint32_t timeout, const char *const str): _name(str)
```
Is that supposed to be a mock?

rajkan01 replied:
Yes,To test mbed watchdog manager implementation we provided duplicate watchdog related implementation to avoid build failure and intention to test mbed watchdog manager stuff. Thanks for your review comment.

0xc0170 commented on drivers/Watchdog.h:
```diff
@@ -46,21 +47,24 @@ namespace mbed {
+     * Assert for multiple calls of start
      */
-    watchdog_status_t start(const uint32_t timeout);
+    void start();
```
why these methods were changed to "void method(...)" ?

rajkan01 replied:
PR#9243 review comments and proposed new design leads to change this to void

donatieng commented on drivers/Watchdog.cpp:
```diff
@@ -20,46 +20,89 @@
+    while (cur_ptr != NULL) {
+        if (cur_ptr == this) {
+            if (cur_ptr == _first) {
+                prev_ptr = _first;
```
`_first` is `this`, can we work with `this` instead?

rajkan01 replied:
No,We just make sure cur_ptr aginst _first node,but tested this code robust enough.

ithinuel replied:
I would agree with @donatieng in the sense that `cur_ptr` and `_first` are variable in this context.
When reading the code, it might be easier to replace usage of `cur_ptr` with `this` in this block.
```c++
// look up for this watchdog in the list to find its predecessor
while ((cur_ptr != this) && (cur_ptr != NULL)) {
    prev_ptr = cur_ptr;
    cur_ptr = cur_ptr->next;
}
if (cur_ptr == NULL) {
    // this was not in the list
    return;
}

if (prev_ptr == NULL) {
    // this was first, set _next as the new _first.
    _first = this->_next;
} else {
    // attach previous watchdog to our next
    prev_ptr->_next = this->_next;
}
// detach us from _next
this->_next = NULL;
_is_initialized = false;
```

donatieng commented on drivers/Watchdog.h:
```diff
@@ -20,21 +20,22 @@
 /** \addtogroup drivers */
 /** A system timer that will reset the system in the case of system failures or
  *  malfunctions.
  *
```
Would be good to explain that this is not driving the HW watchdog directly but can be instantiated multiple times/etc

donatieng commented on drivers/Watchdog.h:
```diff
@@ -71,35 +75,36 @@ class Watchdog {
+    void remove_from_list();
+private:
+    uint32_t _max_timeout; //_max_timeout initialized via constructor while creating instance of this class
+    const char *_name;//To store the details of user
```
Missing spaces before comments

donatieng commented on drivers/Watchdog.h:
```diff
@@ -46,21 +47,24 @@ namespace mbed {
 public:
-    Watchdog() {}
-
+    Watchdog(uint32_t timeout = 1, const char *const str = NULL);
```
Needs doxygen

rajkan01 replied:
Added the doxygen

donatieng commented on platform/mbed_watchdog_mgr.cpp:
```diff
@@ -0,0 +1,112 @@
+ *  its only going to call process to verify all registered users/threads in alive state
+ *
+ */
+mbed::Watchdog watchdog(0, "Platform Watchdog");
```
What's the rationale here?

rajkan01 replied:
To get access of sw watchdog "process" method

ithinuel replied:
Should'nt `Watchdog::process` be a static member then ? That would save a bit of ram.

donatieng commented on UNITTESTS/drivers/Watchdog/test_watchdog.cpp:
```diff
@@ -0,0 +1,142 @@
+    watchdog1.stop();
+    watchdog.stop();
+    watchdog2.stop();
+    EXPECT_EQ(0, TestWatchdog::expect_assert_count);
```
Maybe check reset counter here?

rajkan01 replied:
Yes added the check for reset counter

donatieng reviewed (changes requested):
Hey Rajkumar, thanks - looks good! My main interrogations are around the mock watchdog class which has empty function bodies.

rajkan01 committed:
Incorporated the Reivew Comments

rajkan01 committed:
Incorporated the Reivew Comments

0xc0170 commented on platform/mbed_watchdog_mgr.h:
```diff
@@ -0,0 +1,92 @@
+ * Example:
+ * @code
+ *
+ * mbed_wdog_manager_start();
```
whats the purpose of this code snippet ? only start() invoked, no need to kick or anything else? is it worth having?

rajkan01 replied:
Watchdog has divided into two, HW watchdog & SW watchdog, HW watchdog register for automated periodic call back from lowpowerticker to refresh hw watchdog regathers API used to start the HW watchdog lowpowerticker.

0xc0170 commented on platform/mbed_watchdog_mgr.cpp:
```diff
@@ -0,0 +1,112 @@
+ *  its only going to call process to verify all registered users/threads in alive state
+ *
+ */
+mbed::Watchdog watchdog(0, "Platform Watchdog");
```
why does not follow same as  `get_ticker()` below?

rajkan01 replied:
Watchdog has divided into two, HW watchdog & SW watchdog,HW watchdog intern creates the instance to get access the SW watchdog methods to check states

0xc0170 reviewed (approved):
just 2 questions

ithinuel commented on platform/mbed_watchdog_mgr.h:
```diff
@@ -0,0 +1,92 @@
+ *         stopped, Returns false if the watchdog cannot be disabled
+ *         on the current platform or if the timer was never started.
+ */
+bool mbed_wdog_manager_stop();
```
Shouldn't this method be gated with a `DEVICE_HAS_` (or equivalent) in order to make the user aware of that at compile time ?
This may have critical impact on an application expecting the watchdog to stop.

rajkan01 replied:
I thing but I am not sure,Currently devices have support of Watchdog ,should have support to stop operation ,in case we start add support device only have start then may be we add under macro..

donatieng reviewed (approved):
Looks good to me but I think it would make sense to make `process()` static and not have a "platform" watchdog instance. I know you and @c1728p9 had a chat about it so there might be a rationale I'm missing!

ithinuel commented on platform/mbed_watchdog_mgr.cpp:
```diff
@@ -0,0 +1,112 @@
+ */
+static mbed::LowPowerTicker *get_ticker()
+{
+    static mbed::LowPowerTicker ticker;
```
When is this static instance constructed ?
I wonder if it might not cause some race condition/constructor order issue at some point in the future.
Maybe using a `SingletonPtr<mbed::LowPowerTicker>` would solve this problem.

pan- replied:
@ithinuel `static` variables local to a function are instantiated at first use. `__cxa_guard_acquire`, `__cxa_guard_release` and `__cxa_guard_abort` (see [here](https://github.com/ARMmbed/mbed-os/blob/master/platform/mbed_retarget.cpp#L1436-L1469)) are here to protect the instantiation from race condition. 

You can consider SingletonPtr like _syntactic sugar_; it allows you to write in one line of code what is otherwise written in 5.

rajkan01 replied:
Changed into singletonptr

rajkan01 committed:
Watchdog
 - Changed the process into static method
 - used the singletonptr for creating the low power ticker instance
 - Added the mbed stub into cmake build for cellularnonipsocket,loramacrypto

ithinuel reviewed (approved):
LGTM

rajkan01 committed:
Removed the Watchdog support
 -Removed the watchdog support whichever Board does not have support of LPTICKER

jeromecoutant commented on platform/mbed_watchdog_mgr.cpp:
```diff
@@ -0,0 +1,108 @@
+
+/** Create singleton instance of LowPowerTicker for watchdog periodic call back of kick.
+ */
+static SingletonPtr<mbed::LowPowerTicker> get_ticker()
```
Targets which don't support LP TICKER can't support WATCHDOG API ???

rajkan01 replied:
In the new design watchdog divided into Hw Watcdog and SW watchdog.
1.Hw watchdog registered to low power ticker and expected to call periodically register HW watchdog kick and intern calls HAL Kick API & call process of SW watchdog.(refer mbed_watchdog_mgr.cpp,mbed_wdog_manager_kick() function)
2.SW watchdog has internal mechanism to monitor register service(refer Watchdog.cpp for more info)

fkjagodzinski reviewed (approved):
:+1:

rajkan01 committed:
Revert "Removed the Watchdog support"

This reverts commit ad3c60f908599df0109313e12be53fc410938d7c.

rajkan01 committed:
Added the ticker support
 -Modified the device dont have support of lp ticker will ticker for watchdog callback register

rajkan01 committed:
Fixed the unittest case failure

fkjagodzinski commented on platform/mbed_watchdog_mgr.cpp:
```diff
@@ -0,0 +1,118 @@
+#else
+/** Create singleton instance of Ticker for watchdog periodic call back of kick.
+ */
+static SingletonPtr<mbed::Ticker> get_ticker()
```
```suggestion
static SingletonPtr<mbed::Ticker> _ticker;
```

fkjagodzinski commented on platform/mbed_watchdog_mgr.cpp:
```diff
@@ -0,0 +1,118 @@
+#if DEVICE_LPTICKER
+/** Create singleton instance of LowPowerTicker for watchdog periodic call back of kick.
+ */
+static SingletonPtr<mbed::LowPowerTicker> get_ticker()
```
```suggestion
static SingletonPtr<mbed::LowPowerTicker> _ticker;
```

fkjagodzinski commented on platform/mbed_watchdog_mgr.cpp:
```diff
@@ -0,0 +1,118 @@
+    core_util_critical_section_exit();
+    if (is_watchdog_started) {
+        us_timestamp_t timeout = (MS_TO_US(((elapsed_ms <= 0) ? 1 : elapsed_ms)));
+        get_ticker()->attach_us(mbed::callback(&mbed_wdog_manager_kick), timeout);
```
```suggestion
        _ticker->attach_us(mbed::callback(&mbed_wdog_manager_kick), timeout);
```

fkjagodzinski commented on platform/mbed_watchdog_mgr.cpp:
```diff
@@ -0,0 +1,118 @@
+        if (sts != WATCHDOG_STATUS_OK) {
+            msts = false;
+        } else {
+            get_ticker()->detach();
```
```suggestion
            _ticker->detach();
```

fkjagodzinski reviewed (changes requested):
The current code hard-faults when the ticker callback is executed. I suspect this has to do with `get_ticker()` functions returning `SingletonPtr<mbed::Ticker>` by value, thus copying data from BSS onto the stack. We can fix that and simplify the code by having a global `_ticker` object.

rajkan01 committed:
Fix for Problem copying the instance from BSS to stack
  -Created global instance of either LowPowerTicker or Ticker

fkjagodzinski reviewed (approved):
All correct now. :+1:

fkjagodzinski committed:
Tests: ResetReason: Use the HAL watchdog API

The Watchdog driver has been updated making this test incompatible
with the new API. Use HAL API instead.

rajkan01 committed:
Merge pull request #1 from fkjagodzinski/test_update-watchdog_mgr

Test update watchdog mgr

rajkan01 committed:
Removed the Watchdog and Reset reason for MTB_LAIRD_BL654 & NRF52_DK

rajkan01 committed:
Merge branch 'feature-watchdog-v1-rb' of https://github.com/rajkan01/mbed-os into feature-watchdog-v1-rb

jamesbeyond commented on targets/targets.json:
```diff
@@ -7133,7 +7138,8 @@
             "lf_clock_src": "NRF_LF_SRC_RC"
-        }
+        },
+		"device_has_remove": ["RESET_REASON"]
```
Why not just remove `RESET_REASON` from `MCU_NRF52840` ?

rajkan01 committed:
Removed the watchdog for NRF52

rajkan01 committed:
Revert "Removed the watchdog for NRF52"

This reverts commit 3148e54ec9d6acaba052e4105c013fb209404e16.

rajkan01 committed:
Revert "Removed the Watchdog and Reset reason for MTB_LAIRD_BL654 & NRF52_DK"

This reverts commit c7768e58309eb3e5e822c03c57acc49fcf000a52.

rajkan01 committed:
Removed the RESETREASON on unsupported MCU_NRF52840
